### PR TITLE
Suggest Tests

### PR DESCRIPTION
**Pull request by sokrahta**
finish test suite for soda-fountain suggest resource

[Review on Reviewable](https://reviewable.io/reviews/socrata/soda-fountain/99)

### CODE REVIEW

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala [hunk 1]
```diff
@@ -18,15 +18,16 @@ case class Suggest(datasetDao: DatasetDAO, columnDao: ColumnDAO,
                    httpClient: HttpClient, config: SuggestConfig) {
   val log = org.slf4j.LoggerFactory.getLogger(getClass)
 
-  lazy val spandexAddress = s"${config.host}:${config.port}"
+  if (config.port < 0 || config.port > 65535) throw new IllegalArgumentException("Port out of range (0 to 65535)")
```
Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: Can shorten this to 
```scala
require(config.port >= 0 && config.port <= 65535, "Port out of range (0 to 65535)")
```

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala [hunk 2]
```diff
@@ -53,10 +54,17 @@ case class Suggest(datasetDao: DatasetDAO, columnDao: ColumnDAO,
     }
   }
 
+  // f(dataset name, copy num, column name, text) => spandex uri to get
   def go(req: HttpRequest, resp: HttpServletResponse,
          resourceName: ResourceName, columnName: ColumnName, text: String,
          f: (String, Long, String, String) => URI): Unit = {
-    // f(dataset name, copy num, column name, text)
+    val (ds, cn, col) = internalContext(resp, resourceName, columnName).get
```
Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: If `internalContext(...)` returns `None`, this will just throw an exception and Soda Fountain will return a 500 instead of a 404, won't it?

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala (hunk 1 above)
Comment (sokrahta) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: ah cool concise calm

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala (hunk 2 above)
Comment (sokrahta) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: The 404 response was already committed by `notFound()` inside the for comprehension, verified with tests.
But I'm still not happy with this, it's inelegant and I'm working it out.

Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: Confirmed, when I run this locally and pass in a non-existent dataset ID, Soda Fountain returns a 500 and I see an exception in the log from None.get being called:

```
[23 / GET /suggest/nonexistent/arrestee] INFO NewLoggingHandler 2015-04-07 16:09:51,528 >>> GET /suggest/nonexistent/arrestee
[23 / GET /suggest/nonexistent/arrestee] INFO Suggest 2015-04-07 16:09:51,529 GET /suggest nonexistent :: arrestee [sample]
[23 / GET /suggest/nonexistent/arrestee] INFO Suggest 2015-04-07 16:09:51,530 dataset id not found - nonexistent.arrestee
[23 / GET /suggest/nonexistent/arrestee] ERROR Error 2015-04-07 16:09:51,531 Internal exception: c85b1abf-1d60-4f09-bc77-0ac65a13aafa
java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:313)
	at scala.None$.get(Option.scala:311)
	at com.socrata.soda.server.resources.Suggest.go(Suggest.scala:61)
	at com.socrata.soda.server.resources.Suggest$sampleService$$anonfun$get$1$$anonfun$apply$5.apply(Suggest.scala:105)
	at com.socrata.soda.server.resources.Suggest$sampleService$$anonfun$get$1$$anonfun$apply$5.apply(Suggest.scala:103)
	at com.socrata.soda.server.SodaFountain$$anonfun$1$$anonfun$apply$1.apply(SodaFountain.scala:54)
	at com.socrata.soda.server.SodaFountain$$anonfun$1$$anonfun$apply$1.apply(SodaFountain.scala:52)
	at com.socrata.http.server.util.handlers.NewLoggingHandler$$anonfun$apply$1.apply(NewLoggingHandler.scala:37)
	at com.socrata.http.server.util.handlers.NewLoggingHandler$$anonfun$apply$1.apply(NewLoggingHandler.scala:18)
	at com.socrata.http.server.util.handlers.ThreadRenamingHandler$$anonfun$apply$1.apply(ThreadRenamingHandler.scala:10)
	at com.socrata.http.server.util.handlers.ThreadRenamingHandler$$anonfun$apply$1.apply(ThreadRenamingHandler.scala:6)
	at com.socrata.http.server.FunctionHandler$$anonfun$handle$2.apply(Handlers.scala:26)
	at com.socrata.http.server.FunctionHandler$$anonfun$handle$2.apply(Handlers.scala:16)
	at com.rojoma.simplearm.v2.package$.using(package.scala:41)
	at com.socrata.http.server.FunctionHandler.handle(Handlers.scala:16)
	at org.eclipse.jetty.servlets.gzip.GzipHandler.handle(GzipHandler.java:346)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at com.codahale.metrics.jetty9.InstrumentedHandler.handle(InstrumentedHandler.java:190)
	at com.socrata.http.server.CountingHandler.handle(Handlers.scala:47)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:485)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:290)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:248)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:540)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:606)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:535)
	at java.lang.Thread.run(Thread.java:745)
[23 / GET /suggest/nonexistent/arrestee] INFO Error 2015-04-07 16:09:51,532  responding with error internal-error
[23 / GET /suggest/nonexistent/arrestee] INFO NewLoggingHandler 2015-04-07 16:09:51,532 <<< 3ms ERROR 500
```

Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: The tests only call `internalContext` and not `go`, which is probably overwriting the response code when the `NoSuchElementException` is thrown.

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala [hunk 3]
```diff
@@ -18,15 +18,16 @@ case class Suggest(datasetDao: DatasetDAO, columnDao: ColumnDAO,
...
+    "Connect timeout out of range (milliseconds 0 to 2147483647)")
 
-  private[this] lazy val receiveTimeoutMS = config.receiveTimeout.toMillis.toInt
-  if (receiveTimeoutMS != config.receiveTimeout.toMillis)
-    throw new IllegalArgumentException("Receive timeout out of range (milliseconds must fit in an int)")
+  val receiveTimeoutMS = config.receiveTimeout.toMillis.toInt
+  require(receiveTimeoutMS == config.receiveTimeout.toMillis && receiveTimeoutMS >= 0,
+    "Receive timeout out of range (milliseconds 0 to 2147483647)")
 
   def datasetId(resourceName: ResourceName): Option[String] = {
```
Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: Throw more specific exceptions than Exception (match error?0

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala [hunk 4]
```diff
@@ -53,28 +54,46 @@ case class Suggest(datasetDao: DatasetDAO, columnDao: ColumnDAO,
...
+        val (code, body) = try {
+          getSpandexResponse(f(ds, cn, col, text), req.queryParameters)
+        } catch {
+          case rt: ReceiveTimeout => log.warn(s"Spandex receive timeout $rt")
+            (500, JNull)
+          case ct: ConnectTimeout => log.warn(s"Spandex connect timeout $ct")
+            (500, JNull)
+          case cf: ConnectFailed => log.warn(s"Spandex connect failed $cf")
+            (500, JNull)
+          case e: Exception => log.error(s"Spandex unknown error $e")
```
Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: This is clearly something mysterious so we should not catch it and print a terse log. The stack trace wants to be freeeeee.

File: soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala [hunk 5]
```diff
@@ -53,28 +54,46 @@ case class Suggest(datasetDao: DatasetDAO, columnDao: ColumnDAO,
...
     }
 
     for {
-      ds <- datasetId(resourceName).orElse(notFound(resp, "dataset id"))
-      cn <- copyNum(resourceName).orElse(notFound(resp, "copy"))
-      col <- datacoordinatorColumnId(resourceName, columnName).orElse(notFound(resp, "column"))
+      ds <- datasetId(resourceName).orElse(notFound("dataset id"))
+      cn <- copyNum(resourceName).orElse(notFound("copy"))
+      col <- datacoordinatorColumnId(resourceName, columnName).orElse(notFound("column"))
     } yield {
```
Comment (urmilan) [inline soda-fountain-lib/src/main/scala/com/socrata/soda/server/resources/Suggest.scala]: You don't need the braces here, can just yield the tuple on the same line.

File: soda-fountain-lib/src/test/scala/com/socrata/soda/server/resources/SuggestTest.scala [hunk 6]
```diff
@@ -205,10 +323,11 @@ class SuggestTest extends SpandexTestSuite with Matchers with MockFactory with T
   }
 
   test("spandex receive timeout") {
-    setSpandexResponse(url = "/", body = "receive timeout", syntheticDelayMs = 10000)
+    val path = "/"
+    setSpandexResponse(url = path, body = "receive timeout", syntheticDelayMs = 10000)
     failAfter(6 seconds) {
```
Comment (urmilan) [inline soda-fountain-lib/src/test/scala/com/socrata/soda/server/resources/SuggestTest.scala]: What's the failAfter needed for? Shouldn't adding the synthetic delay just cause the exception to be thrown anyways?

Comment (sokrahta) [inline soda-fountain-lib/src/test/scala/com/socrata/soda/server/resources/SuggestTest.scala]: `failAfter` asserts that the timeout exception is bubbled in a timely manner.

Comment (urmilan) [inline soda-fountain-lib/src/test/scala/com/socrata/soda/server/resources/SuggestTest.scala]: Aaahhhh.
